src/adapter/src/coord/message_handler.rs

```diff
@@ -261,15 +256,44 @@ impl<S: Append + 'static> Coordinator<S> {
             return;
         }
 
-        let plan = match self.plan_statement(&mut session, Statement::CreateSource(stmt), &params) {
-            Ok(Plan::CreateSource(plan)) => plan,
-            Ok(_) => unreachable!("planning CREATE SOURCE must result in a Plan::CreateSource"),
+        let stmts = match result {
+            Ok(ok) => ok,
             Err(e) => return tx.send(Err(e), session),
         };
 
-        let result = self
-            .sequence_create_source(&mut session, plan, depends_on)
-            .await;
+        // First we'll allocate global ids for each source
+        let mut allocation = HashMap::new();
+        let mut allocated_stmts = vec![];
+        for (transient_id, stmt) in stmts {
+            let source_id = match self.catalog.allocate_user_id().await {
+                Ok(id) => id,
+                Err(e) => return tx.send(Err(e.into()), session),
+            };
+            allocation.insert(transient_id, source_id);
+            allocated_stmts.push((source_id, stmt));
+        }
+
+        // Then, we'll rewrite all statements to point to the newly minted global ids and plan them
+        let mut plans = vec![];
+        for (source_id, stmt) in allocated_stmts {
+            let stmt = match mz_sql::names::resolve_transient_ids(&allocation, stmt) {
+                Ok(ok) => ok,
+                Err(e) => return tx.send(Err(e.into()), session),
+            };
+            let depends_on = Vec::from_iter(mz_sql::names::visit_dependencies(&stmt));
+            let plan =
+                match self.plan_statement(&mut session, Statement::CreateSource(stmt), &params) {
```

Comment:
Will planning the create statement ever try and look up the new GlobalId's in the catalog and fail? Just thinking out loud.